server/bleep/src/webserver/aaa.rs

```diff
@@ -0,0 +1,341 @@
+            .as_ref()
+            .expect("instance domain must be provided")
+    );
+
+    let github_oauth_url = &format!(
+        "https://github.com/login/oauth/authorize\
+         ?client_id={client_id}\
+         &state={state}\
+         &allow_signup=false\
+         &redirect_uri={redirect_uri}",
+    );
+
+    auth_layer.initialized_login.insert(state, Instant::now());
+    auth_layer.clean_old_states();
+
+    serde_json::json!({ "oauth_url": github_oauth_url }).to_string()
+}
+
+#[derive(Deserialize)]
+pub(super) struct AuthorizedParams {
+    state: State,
+    code: String,
+}
+
+/// Complete the login flow.
+///
+/// Takes the `state` that has been established previously so we don't leak the actual keys.
+#[utoipa::path(post, path = "/auth/login/complete",
+    responses(
+        (status = 200, description = "Execute query successfully", body = Response),
+        (status = 400, description = "Bad request", body = EndpointError),
+        (status = 500, description = "Server error", body = EndpointError),
+    ),
+)]
+pub(super) async fn authorized(
+    Extension(app): Extension<Application>,
+    Extension(auth_layer): Extension<Arc<AuthLayer>>,
+    Query(params): Query<AuthorizedParams>,
+    jar: PrivateCookieJar,
+) -> impl IntoResponse {
+    let AuthorizedParams { state, code } = params;
+
+    auth_layer
+        .initialized_login
+        .remove(&state)
+        .expect("invalid state key");
+
+    let (client_id, client_secret) = app
+        .config
+        .github_client_id_and_secret()
+        .expect("github client id and secret must be provided");
+
+    let gh_token = auth_layer
+        .client
+        .post(format!(
+            "https://github.com/login/oauth/access_token\
+            ?client_id={client_id}\
+            &client_secret={client_secret}\
+            &code={code}"
+        ))
+        .header("Accept", "application/json")
+        .send()
+        .await
+        .unwrap()
+        .json()
+        .await
+        .unwrap();
+
+    (
+        jar.add(AuthCookie::new(gh_token).to_cookie()),
+        Redirect::to("/"),
+    )
+}
+
+fn unix_time_sec() -> u64 {
+    SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .unwrap()
+        .as_secs()
+}
+
+pub(super) fn router(router: Router, app: Application) -> Router {
+    router
+        .layer(middleware::from_fn_with_state(
+            app,
+            authenticate_authorize_reissue,
+        ))
+        .route("/auth/login/complete", get(authorized))
+        .route("/auth/login/start", get(login))
+        .layer(Extension(Arc::new(AuthLayer::default())))
+}
+
+#[derive(Default)]
+pub(crate) struct AuthLayer {
+    /// Logins that have been initiated, but not completed.
+    ///
+    /// Maps to the time this login attempt was created.
+    initialized_login: DashMap<State, Instant>,
+
+    /// The HTTP client.
+    client: reqwest::Client,
+}
+
+impl AuthLayer {
+    fn clean_old_states(&self) {
```

Comment:
This is not sufficient against DoS protection, as your expiration is time-based, but your memory can be filled up within 5 minutes with login requests.

Reply:
Considering we have to apply rate limiting at the deployment level, this isn't an issue specific to this code. With rate limiting, it's impossible for a single actor to hammer the `state` cache. Multiple actors involves a DDoS, which needs to be mitigated elsewhere as this can flood us at the network level.

Reply:
Unfortunately this assumption doesn't hold in an environment we do not control, such as on premise. A sufficient solution to this DoS would be to limit the number of allowed parallel outstanding login flows to something reasonably large, eg 1024, and cancel anything more than that.

You're correct, DDoS protection can't be in scope here.

Reply:
This is not worth spending time on here. We have the same issues with basic HTTP attacks being able to jam the server because we're using axum without a more robust server in front.